File: cpp/rotting_oranges.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

void bfs(int &c, vector<vector<int> > &grid, queue<pair<int, int> > &q, vector<vector<bool> > &visited) {
    int n = grid.size();
    int m = grid[0].size();
    int k = 0;
    while (!q.empty()) {
        int size = q.size();
        if (k != 0) c++;
        k++;
        for (int s = 0; s < size; s++) {
            pair<int, int> t = q.front();
            q.pop();
            if (t.first < n && t.first >= 0 && t.second + 1 < m && t.second + 1 >= 0 &&
                grid[t.first][t.second + 1] == 1 && !visited[t.first][t.second + 1]) {
                visited[t.first][t.second + 1] = true;
                grid[t.first][t.second + 1] = 2;
                q.push(make_pair(t.first, t.second + 1));
            }
            if (t.first < n && t.first >= 0 && t.second - 1 < m && t.second - 1 >= 0 &&
                grid[t.first][t.second - 1] == 1 && !visited[t.first][t.second - 1]) {
                visited[t.first][t.second - 1] = true;
                grid[t.first][t.second - 1] = 2;
                q.push(make_pair(t.first, t.second - 1));
            }
            if (t.first + 1 < n && t.first + 1 >= 0 && t.second < m && t.second >= 0 &&
                grid[t.first + 1][t.second] == 1 && !visited[t.first + 1][t.second]) {
                visited[t.first + 1][t.second] = true;
                grid[t.first + 1][t.second] = 2;
                q.push(make_pair(t.first + 1, t.second));
            }
            if (t.first - 1 < n && t.first - 1 >= 0 && t.second < m && t.second >= 0 &&
                grid[t.first - 1][t.second] == 1 && !visited[t.first - 1][t.second]) {
                visited[t.first - 1][t.second] = true;
                grid[t.first - 1][t.second] = 2;
                q.push(make_pair(t.first - 1, t.second));
            }
        }
    }
}

int orangesRotting(vector<vector<int> > &grid) {
    int n = grid.size();
    int m = grid[0].size();
    int c = 0;
    queue<pair<int, int> > q;
    vector<vector<bool> > visited(n, vector<bool>(m, false));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (grid[i][j] == 2) {
                visited[i][j] = true;
                q.push(make_pair(i, j));
            }
        }
    }
    bfs(c, grid, q, visited);
    bool flag = false;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (grid[i][j] == 1) {
                flag = true;
                c = -1;
                break;
            }
        }
        if (flag) break;
    }
    return c;
}

int main() {
    vector<vector<int> > v1 = {
            {2, 1, 1},
            {1, 1, 1},
            {0, 1, 2}
    };
    vector<vector<int> > v2 = {
            {0, 2}
    };
    vector<vector<int> > v3 = {
            {2, 1, 1},
            {0, 1, 1},
            {1, 0, 1}
    };
    vector<vector<int> > v4 = {
            {2, 1, 1},
            {1, 1, 0},
            {0, 1, 1}
    };
    vector<vector<int> > v5 = {
            {2, 2},
            {1, 1},
            {0, 0},
            {2, 0}
    };
    cout << orangesRotting(v4);
    return 0;
}

/*
 2, 1, 1
 1, 1, 1
 0, 1, 2

 2, 2, 1
 2, 1, 2
 0, 2, 2
 2, 2, 1
 2, 1, 1
 0, 1, 2

 2, 2, 2
 2, 2, 1
 0, 1, 2
 * */

File: go/tree/rotting_oranges.go
package main

import "fmt"

type Queue[T any] struct {
	data []T
}

func (q *Queue[T]) Empty() bool {
	return q.Len() == 0
}

func (q *Queue[T]) Len() int {
	return len(q.data)
}

func (q *Queue[T]) Push(inp T) {
	q.data = append(q.data, inp)
}

func (q *Queue[T]) Pop() T {
	if q.Len() == 0 {
		panic("queue is empty")
	}
	e := q.data[0]
	q.data = q.data[1:]
	return e
}

func (q *Queue[T]) Front() T {
	if q.Len() == 0 {
		panic("queue is empty")
	}
	return q.data[0]
}

type cell struct {
	x int
	y int
}

func underBoundary(x, y, m, n int) bool {
	return x >= 0 && x < m && y >= 0 && y < n
}

func orangesRotting(grid [][]int) int {
	m := len(grid)
	n := len(grid[0])

	q := Queue[cell]{}
	visited := make(map[cell]struct{})

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == 2 {
				c := cell{x: i, y: j}
				q.Push(c)
				visited[c] = struct{}{}
			}
		}
	}

	ans := 0
	directions := [][]int{
		{1, 0},
		{-1, 0},
		{0, 1},
		{0, -1},
	}

	for !q.Empty() {
		size := q.Len()

		for i := 0; i < size; i++ {
			t := q.Pop()

			for j := 0; j < len(directions); j++ {
				x, y := t.x+directions[j][0], t.y+directions[j][1]
				if underBoundary(x, y, m, n) && grid[x][y] == 1 {
					q.Push(cell{x, y})
					grid[x][y] = 2
				}
			}
		}
		ans++
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == 1 {
				return -1
			}
		}
	}
	if ans == 0 {
		return ans
	}
	return ans - 1
}

func main() {
	fmt.Println(orangesRotting([][]int{{0, 2}}))
}
